File: frontend/types/proposal-version.ts
import { ProposalResponse } from "./proposal";

export interface ProposalVersionResponse {
    id: string;
    proposalId: string;
    version: number;
    action: string;
    comment: string;
    snapshot: ProposalResponse;
    createdAt: string;
    createdBy: string;
    createdByName: string;
}

export const getActionLabel = (action: string): string => {
    switch (action) {
        case "CREATED":
            return "Created";
        case "UPDATED":
            return "Updated";
        case "SENT":
            return "Sent";
        case "ACCEPTED":
            return "Accepted";
        case "REJECTED":
            return "Rejected";
        default:
            return action;
    }
};

export const getActionColor = (action: string): string => {
    switch (action) {
        case "CREATED":
            return "blue";
        case "UPDATED":
            return "amber";
        case "SENT":
            return "indigo";
        case "ACCEPTED":
            return "green";
        case "REJECTED":
            return "red";
        default:
            return "gray";
    }
};

File: backend/src/main/java/com/ultron/backend/domain/entity/ProposalVersion.java
package com.ultron.backend.domain.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;

@Document(collection = "proposal_versions")
@CompoundIndexes({
    @CompoundIndex(name = "proposal_id_version_idx", def = "{'proposalId': 1, 'version': -1}"),
    @CompoundIndex(name = "tenant_proposal_idx", def = "{'tenantId': 1, 'proposalId': 1}")
})
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ProposalVersion {
    @Id
    private String id;

    @Indexed
    private String proposalId; // The ID of the original proposal

    @Indexed
    private String tenantId;

    private Integer version; // Version number (1, 2, 3...)

    // Snapshot of the entire proposal at this point in time
    private Proposal snapshot;

    // Information about the change
    private String action; // CREATED, UPDATED, SENT, ACCEPTED, REJECTED
    private String comment; // Brief description of the change

    // Audit
    private LocalDateTime createdAt;
    private String createdBy;
    private String createdByName;
}
